core/src/machine.rs

```diff
@@ -356,9 +372,82 @@ impl Machine {
     /// 2. Input buffer has been exhausted (all input has been consumed)
     /// 3. Output buffer matches the expected output, as defined by the
     /// [ProgramSpec](ProgramSpec)
+    #[cfg_attr(
+        feature = "wasm",
+        wasm_bindgen(getter, js_name = "isSuccessful")
+    )]
     pub fn is_successful(&self) -> bool {
         self.is_complete()
             && self.input.is_empty()
             && self.output == self.expected_output
     }
+
+    /// A wrapper for [Self::input], to be called from wasm.
+    #[cfg(feature = "wasm")]
+    #[wasm_bindgen(getter, js_name = "input")]
+    pub fn wasm_input(&self) -> Vec<LangValue> {
+        self.input.clone()
+    }
+
+    /// A wrapper for [Self::input], to be called from wasm.
+    #[cfg(feature = "wasm")]
+    #[wasm_bindgen(getter, js_name = "output")]
+    pub fn wasm_output(&self) -> Vec<LangValue> {
+        self.output.clone()
+    }
+
+    /// A wrapper for [Self::registers], to be called from wasm. We can't send
+    /// maps through wasm, so this returns a [JsValue] which is an object
+    /// mapping register names (strings) to their values (`LangValue`).
+    #[cfg(feature = "wasm")]
+    #[wasm_bindgen(getter, js_name = "registers")]
+    pub fn wasm_registers(&self) -> LangValueMap {
+        // Convert the keys of the register map to strings
+        let regs_by_name: HashMap<String, LangValue> = self
+            .registers()
+            .into_iter()
+            .map(|(reg_ref, reg_value)| (reg_ref.to_string(), reg_value))
+            .collect();
+        // Convert the hashmap to a js object. Be careful here!
+        JsValue::from_serde(&regs_by_name).unwrap().unchecked_into()
+    }
```

Comment:
We talked about this a bit but for js-val stuff did you just decide to handle the types in ts instead of adding that custom ts type export?

Reply:
I did add the custom types, look at the return type of the func. These types are defined in `ast::wasm`, and they just map to string version of ts types. So there isn't any actual type checking, but it has the type assigned in TS.